src/gt4py/next/utils.py

```diff
@@ -69,26 +69,99 @@ def flatten_nested_tuple(value: tuple[_T | tuple, ...]) -> tuple[_T, ...]:
         return (value,)
 
 
-def tree_map(fun: Callable[_P, _R]) -> Callable[..., _R | tuple[_R | tuple, ...]]:
+@overload
+def tree_map(fun: Callable[_P, _R], /) -> Callable[..., _R | tuple[_R | tuple, ...]]: ...
+
+
+@overload
+def tree_map(
+    *,
+    collection_type: type | tuple[type, ...],
+    result_collection_constructor: Optional[type] = None,
+) -> Callable[[Callable[_P, _R]], Callable[..., _R | tuple[_R | tuple, ...]]]: ...
+
+
+def tree_map(
+    *args: Callable[_P, _R],
+    collection_type: type | tuple[type, ...] = tuple,
+    result_collection_constructor: Optional[type] = None,  # Todo: check name with Enrique
```

Comment:
```suggestion
    result_collection_constructor: Optional[type] = None,
```
Name is been approved by Hannes.